include/envoy/network/filter.h

```diff
@@ -28,6 +28,46 @@ enum class FilterStatus {
   StopIteration
 };
 
+/**
+ * Callbacks used by individual filter instances to communicate with the filter manager.
+ */
+class NetworkFilterCallbacks {
+public:
+  virtual ~NetworkFilterCallbacks() {}
+
+  /**
+   * @return the connection that owns this filter.
+   */
+  virtual Connection& connection() PURE;
+};
+
+/**
+ * Callbacks used by individual write filter instances to communicate with the filter manager.
+ */
+class WriteFilterCallbacks : public virtual NetworkFilterCallbacks {
+public:
+  virtual ~WriteFilterCallbacks() {}
+
+  /**
+   * Pass data directly to subsequent filters in the filter chain. This method is used in
+   * advanced cases in which a filter needs full control over how subsequent filters view data.
+   * Using this method allows a filter to buffer data (or not) and then periodically inject data
+   * to subsequent filters, indicating end_stream at an appropriate time.
+   * This can be used to implement rate limiting, periodic data emission, etc.
+   *
+   * When using this callback, filters should generally move passed in buffer and return
+   * FilterStatus::StopIteration from their onWrite() call, since use of this method
+   * indicates that a filter does not wish to participate in a standard write flow
+   * and will perform any necessary buffering and continuation on its own.
+   *
+   * @param data supplies the write data to be propagated directly to further filters in the filter
+   *             chain.
+   * @param end_stream supplies the end_stream status to be propagated directly to further filters
+   *                   in the filter chain.
+   */
+  virtual void injectDataToFilterChain(Buffer::Instance& data, bool end_stream) PURE;
```

Comment:
WDYT about `injectWriteDataToFilterChain` or `injectWriteToFilterChain`? So that below we can differentiate with a read version to make directionality more clear?

Reply:
renamed into `injectReadDataToFilterChain()` and `injectWriteDataToFilterChain()`